### PR TITLE
♻️✨Comp backend: refactor progress + allow internal progress bar to use weights

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
**This PR brings some refactoring and new features to the _ProgressBarData_ class:**
- [x] allow usage of sync and async progress reporter for more flexibility
- [x] a progress bar data may be _weighted_, i.e. some steps may have different weights (for example a progress bar of 3 steps with the 1st step being 1/10 of the progress, while the 2nd takes 8/10 and the last step again 1/10
- [x] refactoring --> adds some noise in the PR

**This PR also refactors the main dask-sidecar function:**
- [x] use _ProgressBarData_
- [x] Pulling data and docker image now represents 5% of the progress
- [x] Running the docker image now represents 90% of the progress
- [x] Pushing the data represents 5% of progress

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: packages/service-library/src/servicelib/progress_bar.py [hunk 1]
```diff
@@ -14,13 +16,29 @@ async def __call__(self, progress_value: float) -> None:
...
+    if total == 0:
+        return [1] * steps
+    return [weight / total for weight in weights]
+
+
+_INITIAL_VALUE: Final[float] = -1
+
+
+@dataclass(slots=True, kw_only=True)
 class ProgressBarData:
```
Comment (bisgaard-itis) [inline packages/service-library/src/servicelib/progress_bar.py:36]: That's a cool class! I should start using that!

File: services/dask-sidecar/tests/unit/test_tasks.py [hunk 2]
```diff
@@ -583,12 +583,11 @@ async def test_run_computational_sidecar_dask(
         TaskProgressEvent.parse_raw(msg).progress for msg in progress_sub.buffer
     ]
     # check ordering
-    assert worker_progresses == list(
+    assert worker_progresses == sorted(
```
Comment (bisgaard-itis) [inline services/dask-sidecar/tests/unit/test_tasks.py:586]: Just out of curiosity, why this change?

**Review by bisgaard-itis — APPROVED** [review]
Cool, thanks!

File: services/dask-sidecar/tests/unit/test_tasks.py (hunk 2 above)
Comment (sanderegg) [reply services/dask-sidecar/tests/unit/test_tasks.py:586]: so I found out in my test that
```python
set([0, 0.05, 0.95, 1.0]) == {0, 0.95, 0.05, 1.0}
```

Comment (sanderegg) [reply services/dask-sidecar/tests/unit/test_tasks.py:586]: contrary to c++, python sets are unordered

**Review by GitHK — APPROVED** [review]

File: packages/service-library/src/servicelib/progress_bar.py [hunk 3]
```diff
@@ -70,51 +100,76 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
...
-    async def update(self, value: float = 1) -> None:
+        await self.set_(0)
+
+    def _compute_progress(self, steps: float) -> float:
+        if not self.step_weights:
+            return steps / self.num_steps
+        weight_index = int(steps)
+        return (
+            sum(self.step_weights[:weight_index])
+            + steps % 1 * self.step_weights[weight_index]
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/progress_bar.py:128]: Isn't  ` steps % 1` always be zero? and the mod operation takes precedence over the multiplication

```cmd
>>> 3 % 1
0
>>> 3 % 1 * 2
0
>>> 3 % 1 * 33
0
>>> 10 % 1 * 33
0
>>> 10 % 1 
0
```

File: packages/service-library/src/servicelib/progress_bar.py [hunk 4]
```diff
@@ -70,51 +100,76 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
...
+
+    async def update(self, steps: float = 1) -> None:
+        parent_update_value = 0.0
         async with self._continuous_value_lock:
-            new_progress_value = self._continuous_progress_value + value
-            if new_progress_value > self.steps:
-                new_progress_value = round(new_progress_value)
-            if new_progress_value > self.steps:
+            new_steps_value = self._current_steps + steps
+            if new_steps_value > self.num_steps:
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/progress_bar.py:135]: `if new_steps_value > self.num_steps` is repeated twice
is there a reason to round only if you meet that condition or you can merge both if statements?

File: packages/service-library/src/servicelib/progress_bar.py (hunk 3 above)
Comment (sanderegg) [reply packages/service-library/src/servicelib/progress_bar.py:128]: no because in case that progress bar has children, the steps value being a float, there is a modulo result.
and it is correctly first modulo then multiply.

File: packages/service-library/src/servicelib/progress_bar.py (hunk 4 above)
Comment (sanderegg) [reply packages/service-library/src/servicelib/progress_bar.py:135]: yes, I had to think a gain, but the reason why is:
- first there is a check (since it is a float value it can be slightly above the max value), therefore it is rounded
- if it is still above the value then comes the warning

**Review by pcrespov — APPROVED** [review]

**Review by matusdrobuliak66 — APPROVED** [review]
